tool/modules/doip.py

```diff
@@ -0,0 +1,973 @@
+from lib.iso14229_1 import Constants, NegativeResponseCodes
+from doipclient import DoIPClient
+from doipclient.connectors import DoIPClientUDSConnector
+from udsoncan.client import Client
+from udsoncan.services import *
+from sys import stdout, version_info
+import argparse
+import datetime
+import time
+import sys
+import struct
+
+
+# Handle large ranges efficiently in both python 2 and 3
+if version_info[0] == 2:
+    range = xrange
+
+UDS_SERVICE_NAMES = {
+    0x10: "DIAGNOSTIC_SESSION_CONTROL",
+    0x11: "ECU_RESET",
+    0x14: "CLEAR_DIAGNOSTIC_INFORMATION",
+    0x19: "READ_DTC_INFORMATION",
+    0x20: "RETURN_TO_NORMAL",
+    0x22: "READ_DATA_BY_IDENTIFIER",
+    0x23: "READ_MEMORY_BY_ADDRESS",
+    0x24: "READ_SCALING_DATA_BY_IDENTIFIER",
+    0x27: "SECURITY_ACCESS",
+    0x28: "COMMUNICATION_CONTROL",
+    0x2A: "READ_DATA_BY_PERIODIC_IDENTIFIER",
+    0x2C: "DYNAMICALLY_DEFINE_DATA_IDENTIFIER",
+    0x2D: "DEFINE_PID_BY_MEMORY_ADDRESS",
+    0x2E: "WRITE_DATA_BY_IDENTIFIER",
+    0x2F: "INPUT_OUTPUT_CONTROL_BY_IDENTIFIER",
+    0x31: "ROUTINE_CONTROL",
+    0x34: "REQUEST_DOWNLOAD",
+    0x35: "REQUEST_UPLOAD",
+    0x36: "TRANSFER_DATA",
+    0x37: "REQUEST_TRANSFER_EXIT",
+    0x38: "REQUEST_FILE_TRANSFER",
+    0x3D: "WRITE_MEMORY_BY_ADDRESS",
+    0x3E: "TESTER_PRESENT",
+    0x7F: "NEGATIVE_RESPONSE",
+    0x83: "ACCESS_TIMING_PARAMETER",
+    0x84: "SECURED_DATA_TRANSMISSION",
+    0x85: "CONTROL_DTC_SETTING",
+    0x86: "RESPONSE_ON_EVENT",
+    0x87: "LINK_CONTROL"
+}
+
+DELAY_DISCOVERY = 0.2
+DELAY_TESTER_PRESENT = 0.5
+DELAY_SECSEED_RESET = 0.01
+DELAY_FUZZ_RESET = 3.901
+TIMEOUT_SERVICES = 0.2
+
+# Max number of arbitration IDs to backtrack during verification
+VERIFICATION_BACKTRACK = 5
+# Extra time in seconds to wait for responses during verification
+VERIFICATION_EXTRA_DELAY = 0.5
+
+BYTE_MIN = 0x00
+BYTE_MAX = 0xFF
+
+DUMP_DID_MIN = 0x0000
+DUMP_DID_MAX = 0xFFFF
+DUMP_DID_TIMEOUT = 0.2
+
+class DevNull:
+    # Supress errors solution: 
+    # https://stackoverflow.com/questions/5925918/python-suppressing-errors-from-going-to-commandline
+    
+    def write(self, msg):
+        pass
+
+# Duplicate testing from https://www.iditect.com/guide/python/python_howto_find_the_duplicates_in_a_list.html
+def find_duplicates(sequence):
+    first_seen = set()
+    first_seen_add = first_seen.add
+    duplicates = set(i for i in sequence if i in first_seen or first_seen_add(i))
+    return duplicates
+
+def ecu_reset(client, reset_type):
+    if reset_type == 1:
+        client.ecu_reset(ECUReset.ResetType.hardReset)
+    elif reset_type == 2:
+        client.ecu_reset(ECUReset.ResetType.keyOffOnReset)
+    elif reset_type == 3:
+        client.ecu_reset(ECUReset.ResetType.softReset)
+    elif reset_type == 4:
+        client.ecu_reset(ECUReset.ResetType.enableRapidPowerShutDown)
+    elif reset_type == 5:
+        client.ecu_reset(ECUReset.ResetType.disableRapidPowerShutDown)
+
+def extended_session(client, session_type):
```

Comment:
Basically I implemented it because there are a couple of places where the user can specify an option for the diagnostic session he/she needs (e.g. 319, 514, 632, etc.). This means that this exact code of the extended_session() should be copied more than one time inside the code which from my perspective makes the code less clean. Let me know your preference and I will resolve it :)

Reply:
I see - let's keep it like this, then! Thanks for explaining 👍